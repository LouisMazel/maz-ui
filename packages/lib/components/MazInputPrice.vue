<template>
  <MazInput
    v-model="displayPrice"
    class="maz-input-price"
    v-bind="$attrs"
    @focus="isActive = true"
    @blur="isActive = false"
  >
    <template #left-icon>
      <slot v-if="!noIcon" name="left-icon">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
          aria-hidden="true"
          class="maz-text-mute maz-ml-1 maz-h-6 maz-w-6"
        >
          <path
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M17 9V7a2 2 0 00-2-2H5a2 2 0 00-2 2v6a2 2 0 002 2h2m2 4h10a2 2 0 002-2v-6a2 2 0 00-2-2H9a2 2 0 00-2 2v6a2 2 0 002 2zm7-5a2 2 0 11-4 0 2 2 0 014 0z"
          />
        </svg>
      </slot>
    </template>
  </MazInput>
</template>

<script lang="ts" setup>
  import { computed, nextTick, onBeforeMount, ref } from 'vue'
  import { currency as currencyFilter } from '../modules'
  import MazInput from './MazInput.vue'

  const props = defineProps({
    modelValue: { type: [Number, String], default: undefined },
    currency: { type: String, default: 'EUR' },
    locale: { type: String, default: 'fr-FR' },
    min: { type: Number, default: 0 },
    max: { type: Number, default: Number.POSITIVE_INFINITY },
    noIcon: { type: Boolean, default: false },
  })

  const emits = defineEmits(['update:model-value', 'formatted'])

  const isActive = ref(false)
  const valueString = computed<string>(() => {
    return getAdjustedPrice(props.modelValue).toString()
  })
  const valueNumber = computed<number>(() => {
    const value = typeof props.modelValue === 'string' ? Number(props.modelValue) : props.modelValue

    return getAdjustedPrice(value)
  })

  const priceFormatted = computed(() =>
    currencyFilter(valueNumber.value, props.locale, { currency: props.currency }),
  )

  const getAdjustedPrice = (value?: string | number) => {
    let newValue =
      typeof value === 'string'
        ? Number.parseFloat(
            // eslint-disable-next-line no-useless-escape
            value.replace(',', '.').replaceAll(/[^\d.]/g, ''),
          )
        : value
    if (!newValue || Number.isNaN(newValue)) newValue = 0
    if (newValue < props.min) newValue = props.min
    if (newValue > props.max) newValue = props.max

    return newValue
  }

  const displayPrice = computed({
    get: () => {
      if (isActive.value) return valueString.value
      if (typeof props.modelValue === 'number') return priceFormatted.value
      return undefined
    },
    set: (value) => {
      if (Number.isNaN(value)) {
        emitValues()
      } else {
        const adjustedPrice = getAdjustedPrice(value)
        emitValues(adjustedPrice)
      }
    },
  })

  const emitValues = async (newValue?: number) => {
    const adjustedPrice = typeof newValue === 'number' ? getAdjustedPrice(newValue) : undefined
    emits('update:model-value', adjustedPrice)

    await nextTick()
    emits('formatted', priceFormatted.value)
  }

  onBeforeMount(() => {
    emitValues(getAdjustedPrice(props.modelValue))
  })
</script>
