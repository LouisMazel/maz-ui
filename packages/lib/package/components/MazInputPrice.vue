<template>
  <MazInput
    v-model="displayPrice"
    class="maz-input-price"
    v-bind="$attrs"
    left-icon="cash"
    @focus="isActive = true"
    @blur="isActive = false"
  />
</template>

<script lang="ts" setup>
  import { computed, nextTick, onBeforeMount, ref } from 'vue'
  import { currency } from '../filters'
  import MazInput from './MazInput.vue'

  const props = defineProps({
    modelValue: { type: [Number, String], default: undefined },
    currency: { type: String, default: 'EUR' },
    locale: { type: String, default: 'fr-FR' },
    min: { type: Number, default: 0 },
    max: { type: Number, default: Infinity },
  })

  const emits = defineEmits(['update:model-value', 'formatted'])

  const isActive = ref(false)
  const valueString = computed<string>(() => {
    return typeof props.modelValue === 'number'
      ? getAdjustedPrice(props.modelValue).toString()
      : getAdjustedPrice(props.modelValue).toString()
  })
  const valueNumber = computed<number>(() => {
    const value =
      typeof props.modelValue === 'string'
        ? Number(props.modelValue)
        : props.modelValue

    return getAdjustedPrice(value)
  })

  const priceFormatted = computed(() =>
    currency(valueNumber.value, props.locale, { currency: props.currency }),
  )

  const getAdjustedPrice = (value?: string | number) => {
    let newValue =
      typeof value === 'string'
        ? parseFloat(
            // eslint-disable-next-line no-useless-escape
            value.replace(',', '.').replace(/[^\d\.]/g, ''),
          )
        : value
    if (!newValue || isNaN(newValue)) newValue = 0
    if (newValue < props.min) newValue = props.min
    if (newValue > props.max) newValue = props.max

    return newValue
  }

  const displayPrice = computed({
    get: () => {
      if (isActive.value) return valueString.value
      if (props.modelValue) return priceFormatted.value
      return undefined
    },
    set: (value) => {
      if (!value) {
        emitValues(undefined)
      } else {
        const adjustedPrice = getAdjustedPrice(value)
        emitValues(adjustedPrice)
      }
    },
  })

  const emitValues = async (newValue?: number) => {
    const adjustedPrice = newValue ? getAdjustedPrice(newValue) : undefined
    emits('update:model-value', adjustedPrice)

    await nextTick()
    emits('formatted', priceFormatted.value)
  }

  onBeforeMount(() => {
    emitValues(getAdjustedPrice(props.modelValue))
  })
</script>
