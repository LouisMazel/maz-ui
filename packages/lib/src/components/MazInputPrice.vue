<script lang="ts" setup>
import MazInput from '@components/MazInput.vue'
import { currency as currencyFilter, type FilterCurrencyOptions } from '@filters/currency'
import { computed, defineAsyncComponent, nextTick, onMounted, ref, watch } from 'vue'

const props = withDefaults(defineProps<MazInputPriceProps>(), {
  modelValue: undefined,
  currency: 'EUR',
  locale: 'fr-FR',
  min: Number.NEGATIVE_INFINITY,
  max: Number.POSITIVE_INFINITY,
  noIcon: false,
})

const emits = defineEmits<{
  'update:model-value': [value: number | undefined]
  'input': [value: number | undefined]
  'formatted': [value: string | undefined]
  'blur': [value: Event]
  'focus': [value: Event]
}>()

const BanknotesIcon = defineAsyncComponent(() => import('@icons/banknotes.svg'))

export interface MazInputPriceProps {
  /** @model The value of the input */
  modelValue?: number
  /** The currency to use */
  currency?: string
  /** The locale to use */
  locale?: string
  /** The minimum value that the input can accept */
  min?: number
  /** The maximum value that the input can accept */
  max?: number
  /** The input will be displayed without icon */
  noIcon?: boolean
  /** The input will be displayed in full width */
  block?: boolean
  /** Will display the input in error state. */
  error?: boolean
  /** The hint text to display below the input. */
  hint?: string
  /** Will display the input in success state. */
  success?: boolean
  /** Will display the input in warning state. */
  warning?: boolean
  /** Options for the currency helper - [see](https://maz-ui.com/helpers/currency) */
  currencyOptions?: Omit<FilterCurrencyOptions, 'currency'>
}

const internalValue = ref<number | undefined>(getAdjustedPrice(props.modelValue))
watch(() => props.modelValue, updateInternalValue)

const isActive = ref(false)
const valueString = computed<string | undefined>(() => {
  return props.modelValue?.toString()
})
const valueNumber = computed<number | undefined>(() => {
  return props.modelValue
})

const priceFormatted = computed(() =>
  typeof valueNumber.value === 'number' ? currencyFilter(valueNumber.value, props.locale, { ...props.currencyOptions, currency: props.currency }) : undefined,
)

function getAdjustedPrice(value?: string | number) {
  const isNegative = typeof value === 'string' && value.startsWith('-')

  let newValue = typeof value === 'string'
    ? Number.parseFloat(value.replace(',', '.').replaceAll(/[^\d.]/g, ''))
    : value

  newValue = isNegative && newValue ? -newValue : newValue

  if (typeof newValue !== 'number' || Number.isNaN(newValue))
    newValue = undefined
  if (newValue && newValue < props.min)
    newValue = props.min
  if (newValue && newValue > props.max)
    newValue = props.max

  return newValue
}

const displayPrice = computed(() => {
  if (isActive.value)
    return valueString.value
  if (typeof props.modelValue === 'number')
    return priceFormatted.value

  return undefined
})

async function emitValues(newValue?: string | number) {
  const adjustedPrice = getAdjustedPrice(newValue)
  emits('update:model-value', adjustedPrice)

  await nextTick()
  emits('formatted', priceFormatted.value)
}

onMounted(() => {
  emitValues(props.modelValue)
})

function onBlur(event: Event) {
  isActive.value = false
  emitValues(internalValue.value)
  emits('blur', event)
}
function onFocus(event: Event) {
  isActive.value = true
  emits('focus', event)
}
function onInput() {
  emits('input', internalValue.value)
}
function updateInternalValue(value?: string | number) {
  internalValue.value = getAdjustedPrice(value)
}
</script>

<template>
  <MazInput
    :model-value="displayPrice"
    class="maz-input-price m-reset-css"
    :block
    :error
    :success
    :warning
    :hint
    @focus="onFocus"
    @blur="onBlur"
    @input="onInput"
    @keydown.enter="emitValues(internalValue)"
    @update:model-value="updateInternalValue($event)"
  >
    <template #left-icon>
      <slot v-if="!noIcon" name="left-icon">
        <BanknotesIcon class="maz-text-xl" />
      </slot>
    </template>
  </MazInput>
</template>
