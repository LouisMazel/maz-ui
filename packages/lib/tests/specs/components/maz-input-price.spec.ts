import MazInputPrice from '@components/MazInputPrice.vue'
import { shallowMount } from '@vue/test-utils'

describe('components/MazInput.vue', () => {
  expect(MazInputPrice).toBeTruthy()

  const wrapper = shallowMount(MazInputPrice, {
    props: {
      modelValue: 12,
    },
  })

  it('should have the model value', () => {
    expect(wrapper.vm.modelValue).toBe(12)
  })

  it('should have the formatted price displayed', () => {
    // @ts-expect-error - test case
    expect(wrapper.vm.displayPrice).toBe('12,00 €')
  })

  it('should have the formatted price displayed with zero value', async () => {
    wrapper.setProps({
      modelValue: 0,
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.modelValue).toBe(0)
    // @ts-expect-error - test case
    expect(wrapper.vm.displayPrice).toBe('0,00 €')
  })
})
