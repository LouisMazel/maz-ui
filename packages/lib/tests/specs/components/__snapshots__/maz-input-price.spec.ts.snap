// Vitest Snapshot v1

exports[`components/MazInput.vue > Should match with the snapshot 1`] = `
"<div data-v-562a2cff=\\"\\" class=\\"m-input maz-input-price --primary --md\\">
  <div data-v-562a2cff=\\"\\" class=\\"m-input-wrapper --default-border maz-rounded\\">
    <div data-v-562a2cff=\\"\\" class=\\"m-input-wrapper-left\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\" class=\\"maz-text-mute maz-ml-1 maz-h-6 maz-w-6\\">
        <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M17 9V7a2 2 0 00-2-2H5a2 2 0 00-2 2v6a2 2 0 002 2h2m2 4h10a2 2 0 002-2v-6a2 2 0 00-2-2H9a2 2 0 00-2 2v6a2 2 0 002 2zm7-5a2 2 0 11-4 0 2 2 0 014 0z\\"></path>
      </svg></div>
    <div data-v-562a2cff=\\"\\" class=\\"m-input-wrapper-input\\"><input data-v-562a2cff=\\"\\" id=\\"MazInput-2\\" type=\\"text\\" name=\\"input\\" class=\\"maz-input-price m-input-input\\">
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
</div>"
`;
